theano/compile/function_module.py

```diff
@@ -496,6 +500,65 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory=False):
+        """
+        Copy this function. Copied function will have separated maker and
+        fgraph with original function. User can choose whether to separate
+        storage by changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two
+            function share intermediate storages(storages except input and
+            output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
+
+            # copy fgraph and get memo
+            maker = self.maker
+            fg_cpy, memo = maker.fgraph.clone_get_equiv(attach_feature=False)
+
+            # construct new storage_map that map new variable to old storage
+            # so that the ensuing function shares storage with the original one
+            new_storage_map = {}
+            storage_map = self.fn.storage_map
+
+            # TODO: We could share the output storage, but we must make sure
+            # 2 different function call won't override each other values. This
+            # is already done elsewhere, so to reuse it the user would need to
+            # use Out(var, borrow=True) and maybe the mutable=True flag too.
+            # But to be safe for now as it isn't documented and we aren't sure
+            # it is well tested, we don't share the part of the storage_map.
+            for key in storage_map.keys():
+                if key not in maker.fgraph.outputs:
+                    new_storage_map[memo[key]] = storage_map[key]
+
+            # copy input storages if it's mutable
+            input_storage = []
+            for in_ori, in_cpy in zip(maker.inputs, ins):
+                is_const = isinstance(in_ori.variable, theano.tensor.Constant)
+                if is_const or not in_ori.mutable:
+                    storage = getattr(in_ori, 'value', None)
+                    in_cpy.value = in_ori.value
```

Comment:
I agree that if it is a Constant, we should not copy the data. But I think we should modify the class Constant, by adding to it a method `__deepcopy__` that do not copy the data. This would remove that line and would be cleaner. The class Container in the file link.py implement it to work around a numpy problem with deepcopy. Maybe it can help you.